File: leetcode-04/solutions/cpp/0202.cpp
class Solution {
 public:
  bool isHappy(int n) {
    int slow = helper(n);
    int fast = helper(helper(n));

    while (slow != fast) {
      slow = helper(slow);
      fast = helper(helper(fast));
    }

    if (slow == 1) return true;
    return false;
  }

 private:
  int helper(int n) {
    int sum = 0;

    while (n) {
      sum += pow(n % 10, 2);
      n /= 10;
    }

    return sum;
  }
};

File: leetcode-04/solutions/python3/0202.py
class Solution:
    def isHappy(self, n: int) -> bool:
        slow = self.helper(n)
        fast = self.helper(self.helper(n))

        while slow != fast:
            slow = self.helper(slow)
            fast = self.helper(self.helper(fast))

        if slow == 1:
            return True
        return False

    def helper(self, n: int) -> bool:
        sum = 0

        while n:
            sum += pow(n % 10, 2)
            n //= 10

        return sum
